### PR TITLE
feat: monitor parent process and terminate if parent process is no longer running

### PR DESCRIPTION
**Pull request by philasmar**
*Issue #, if available:*
DOTNET-5083

*Description of changes:*
feat: monitor parent process and terminate if parent process is no longer running

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs [hunk 1]
```diff
@@ -35,7 +33,58 @@ public async Task ExecuteAsync(CancellationToken cancellationToken = default(Can
...
+                var webTask = host.RunAsync(cancellationToken);
+
+                //This call will wait until one of the tasks completes.
+                //The monitor task will complete if a parent pid is not found.
+                await Task.WhenAny(monitorTask, webTask);
+
+                //The monitor task is expected to complete only when a parent pid is not found.
+                if (monitorTask.IsCompleted && monitorTask.Result?.ProcessNotFound == true)
+                {
+                    throw new ParentProcessNotFoundException("The parent process is no longer running.");
```
Comment (normj) [inline src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:52]: I don't think we have to throw an exception causing a non-zero exit code. I think we should just log that the parent process exited so we are shutting down server mode.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs (hunk 1 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs]: Fixed.

**Review by normj — APPROVED** [review]

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs [hunk 2]
```diff
@@ -91,5 +120,29 @@ private IEncryptionProvider CreateEncryptionProvider()
...
+            while (true)
+            {
+                try
+                {
+                    Process.GetProcessById((int)_parentPid);
+                    await Task.Delay(1000, token);
+                }
+                catch
+                {
+                    return new MonitoringExited { ProcessNotFound = true };
```
Comment (ganeshnj) [inline src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:143]: Is there a specific exception that you can catch?

Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:143]: There is, but the method throws multiple exceptions. I want to consider any issue retrieving the parent process as the same issue for our context. If anything goes wrong, then the parent process is not available.

Comment (ganeshnj) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:143]: Then I think we don't need

```
        public class MonitoringExited
        {
            public bool ProcessNotFound {get;set;}
        }
```
Setting ProcessNotFound to true gives a false indication. We should simply return a boolean that indicates the whether GetProcessById threw an exception or not.

If needed we can always expand return type but I don't reason for MonitoringExited existence.

Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs]: I removed MonitoringExited and returned a bool.

**Pull request merged**